core/src/main/java/com/venia/core/models/commerce/services/ClearFullDispatcherCacheBasedOnCategory.java

```diff
@@ -0,0 +1,248 @@
+        String[] categoryUids = extractCategoryUidsFromContext(context);
+        if (!isValidCategoryUids(categoryUids)) {
+            return Collections.emptyList();
+        }
+
+        // Fetch category data and determine if cache invalidation is needed
+        List<Map<String, Object>> categories = fetchCategoryData(context, categoryUids);
+        if (categories.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        if (shouldInvalidateFullCache(categories, navigationStructureDepth)) {
+            return Collections.singletonList(context.getStorePath());
+        }
+
+        return Collections.emptyList();
+    }
+
+    /**
+     * Extracts category UIDs from the context's attribute data.
+     *
+     * @param context The cache invalidation context
+     * @return Array of category UIDs
+     */
+    private String[] extractCategoryUidsFromContext(DispatcherCacheInvalidationContext context) {
+        if (context == null) {
+            LOGGER.warn("Context is null when extracting category UIDs");
+            return new String[0];
+        }
+
+        List<String> attributeData = context.getAttributeData();
+        return attributeData != null ? attributeData.toArray(new String[0]) : new String[0];
+    }
+
+    /**
+     * Fetches category information using GraphQL query.
+     *
+     * @param context The cache invalidation context
+     * @param categoryUids Array of category UIDs to query
+     * @return List of category data maps
+     */
+    private List<Map<String, Object>> fetchCategoryData(DispatcherCacheInvalidationContext context, String[] categoryUids) {
+        if (context == null || categoryUids == null) {
+            LOGGER.warn("Invalid parameters for fetching category data");
+            return Collections.emptyList();
+        }
+
+        String query = buildCategoryQuery(categoryUids);
+        if (query == null) {
+            LOGGER.warn("Failed to build category query");
+            return Collections.emptyList();
+        }
+
+        Query data = getGraphqlResponseData(context.getGraphqlClient(), query);
+        if (data == null) {
+            return Collections.emptyList();
+        }
+
+        List<CategoryTree> categories = data.getCategoryList();
+        if (categories == null || categories.isEmpty()) {
+            LOGGER.debug("No categories found for UIDs: {}", (Object) categoryUids);
+            return Collections.emptyList();
+        }
+
+        return categories.stream()
+                .map(category -> {
+                    Map<String, Object> map = new HashMap<>();
+                    map.put("uid", category.getUid());
+                    map.put(LEVEL_KEY, category.getLevel());
+                    return map;
+                })
+                .collect(Collectors.toList());
+    }
+
+    protected Query getGraphqlResponseData(MagentoGraphqlClient client, String query) {
+        if (client == null || query == null) {
```

Comment:
Again client should be guaranteed to be non-null. The query is put together by the customer code, why would they pass a null query to this method?